File: FileTreeWalker.hs
module FileTreeWalker where

import System.Directory (doesDirectoryExist, getDirectoryContents)
import System.FilePath ((</>))
import qualified Data.Text.Lazy.IO as T
import qualified Data.Text.Lazy as T


walkFileTree :: FilePath -> (FilePath -> IO ()) -> IO ()
walkFileTree f proc = do
  isDirectory <- doesDirectoryExist f
  if isDirectory
    then do
      files <- getDirectoryContents f
      mapM_ ((`walkFileTree` proc) . (f </>)) $ filter (`notElem` [".", ".."]) files
    else proc f

main :: IO ()
main = walkFileTree "C:\\" $ T.putStrLn . T.pack

File: file_tree_walker.ml
module FileTreeWalker : sig
  val walk_file_tree : string -> (string -> unit) -> unit
end = struct
  let rec walk_file_tree f proc =
    if Sys.is_directory f
      then
        let files = Sys.readdir f in
        Array.iter (fun child -> walk_file_tree (Filename.concat f child) proc) files
      else proc f
end

let () =
  FileTreeWalker.walk_file_tree "C:\\" print_endline
